content/cucumber/debugging.md

```diff
@@ -0,0 +1,157 @@
+    @scenario_name = scenario.scenario_outline.name
+  end
+  Rails.logger.info("[Cucumber] starting the #{@scenario_name}")
+end
+
+# `STEP=1 cucumber` to pause after each step
+AfterStep do |scenario|
+  next unless ENV['STEP']
+  unless defined?(@counter)
+    puts "Stepping through #{@scenario_name}"
+    @counter = 0
+  end
+  @counter += 1
+  print "After step ##{@counter}/#{scenario.send(:steps).try(:count)}: "\
+        "#{scenario.send(:steps).to_a[@counter].try(:name) ||
+        '[RETURN to continue]'}..."
+  STDIN.getc
+end
+
+AfterStep do |scenario|
+  CucumberCounters.step_counter += 1
+  step = CucumberCounters.step_counter
+  file_name = format('tmp/capybara/step_%03d.png', step)
+  Rails.logger.info("[Cucumber] after step: #{@scenario_name}, step: #{step}")
+  next unless scenario.source_tag_names.include?('@intermittent')
+  begin
+    Capybara.page.save_screenshot(file_name, full: true)
+    Rails.logger.info("[Cucumber] Screenshot #{step} saved")
+  rescue
+    Rails.logger.info("[Cucumber] Can not make screenshot of #{step}")
+  end
+end
+
+AfterStep do
+  begin
+    execute_script "$(window).unbind('beforeunload')"
+  rescue => e
+    Rails.logger.error("An error was encountered and rescued")
+    Rails.logger.error(e.backtrace)
+  end
+end
+
+def dismiss_nav_warning
+  execute_script "$(window).unbind('beforeunload')"
+  wait_until_jquery_inactive
+end
+
+def wait_until_jquery_inactive
+  Capybara.using_wait_time(Capybara.default_max_wait_time) do
+    page.evaluate_script('jQuery.active').zero?
+  end
+end
+```
+By setting an environment variable, you can cause Cucumber to use various debugging tools, and you can combine them by setting multiple environment variables.
+{{% /block %}}
+
+{{% block "java" %}}
+Here is how to debug your scenarios on the JVM, by stepping through the steps of each scenario:
+
+1. Set a conditional breakpoint in the method `cucumber.runtime.Utils#invoke`, at the line `return targetMethod.invoke(target, args)` (line 40 in `cucumber-jvm` v1.2.5) and specify the following snippet as the condition (see also [https://www.jetbrains.com/help/idea/2017.1/breakpoints-2.html](https://www.jetbrains.com/help/idea/2017.1/breakpoints-2.html) for more details on how to set one on IDEA):
+
+        Package pkg = target.getClass().getPackage();
+        if (pkg == null) {
+            return false;
+        }
+        return !target.getClass().getPackage().getName().startsWith("cucumber");
+ 
+2. Run your [RunCukesTest](https://github.com/cucumber/cucumber-java-skeleton/blob/master/src/test/java/skeleton/RunCukesTest.java) in debug mode
+3. Assuming you haven't set any other breakpoints, the execution will stop at `Utils#invoke`
+4. Now you can either:
+   - _Step into_ to start debugging the method implementing the first step of the scenario
+   - Or _Resume_ the execution to run the current step and jump to the next one
+5. And so on..
+{{% /block %}}
+
+{{% block "javascript" %}}
+TBD.
```

Comment:
Anyone available to add some JavaScript debugging tipcs? (note to self: check the Cucumber-js documentation)